File: combat/TurnManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lanes.Combat.Units;

namespace Lanes.Combat;

/// <summary>
/// Determines and advances the turn order for a group of units.
/// Turn order is resolved by descending Speed; ties are broken by the
/// unit's position in the roster (lower index acts first).
///
/// The manager does not mutate units — it only reads their stats.
/// </summary>
public sealed class TurnManager
{
    private readonly List<UnitInstance> _order = new();
    private int _currentIndex;

    /// <summary>
    /// Rebuilds the turn order from <paramref name="units"/>.
    /// Call at the start of each round after units may have died or had Speed changed.
    /// </summary>
    public void SetOrder(IEnumerable<UnitInstance> units)
    {
        if (units is null) throw new ArgumentNullException(nameof(units));

        _order.Clear();
        _order.AddRange(
            units
                .Where(u => u.IsAlive)
                .OrderByDescending(u => u.Stats.Speed));

        _currentIndex = 0;
    }

    /// <summary>Ordered snapshot of living units for this round (highest Speed first).</summary>
    public IReadOnlyList<UnitInstance> Order => _order.AsReadOnly();

    /// <summary>The unit whose turn it currently is, or <c>null</c> if the round is over.</summary>
    public UnitInstance? Current =>
        _currentIndex < _order.Count ? _order[_currentIndex] : null;

    /// <summary>
    /// Advances to the next living unit in turn order.
    /// Dead units are skipped automatically.
    /// </summary>
    /// <returns>
    /// The unit that is now acting, or <c>null</c> when the round is complete.
    /// </returns>
    public UnitInstance? Advance()
    {
        _currentIndex++;
        // Skip over units that died mid-round.
        while (_currentIndex < _order.Count && !_order[_currentIndex].IsAlive)
            _currentIndex++;

        return Current;
    }

    /// <summary><c>true</c> when all units in the current round have taken their turn.</summary>
    public bool IsRoundComplete => _currentIndex >= _order.Count;
}

File: src/app/core/engine/turn-manager.ts
import { UnitInstance } from '../models/unit-instance.model';

/**
 * Determines and advances the turn order for units on the battlefield.
 * Order is resolved by descending Speed. Dead units are skipped.
 */
export class TurnManager {
  private readonly _order: UnitInstance[] = [];
  private _currentIndex = 0;
  private _roundNumber = 1;

  public setOrder(units: readonly UnitInstance[]): void {
    this._order.length = 0;
    const living = units.filter((u) => u.isAlive).sort((a, b) => b.stats.speed - a.stats.speed);
    this._order.push(...living);
    this._currentIndex = 0;
  }

  public get order(): readonly UnitInstance[] {
    return this._order;
  }

  public get currentIndex(): number {
    return this._currentIndex;
  }

  public get roundNumber(): number {
    return this._roundNumber;
  }

  public get current(): UnitInstance | null {
    return this._currentIndex < this._order.length ? this._order[this._currentIndex] : null;
  }

  public advance(): UnitInstance | null {
    this._currentIndex++;
    while (this._currentIndex < this._order.length && !this._order[this._currentIndex].isAlive) {
      this._currentIndex++;
    }

    return this.current;
  }

  public startNewRound(units: readonly UnitInstance[]): void {
    this._roundNumber++;
    this.setOrder(units);
  }

  public get isRoundComplete(): boolean {
    return this._currentIndex >= this._order.length;
  }
}
